adapters/amqp-vertx-base/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -754,6 +757,22 @@ protected void handleRemoteSenderOpenForCommands(
         });
     }
 
+    private void closeConnectionOnTerminalError(final Throwable error, final ProtonConnection conn,
+            final AmqpContext ctx, final Span span) {
+        final ResourceIdentifier address = ctx.getAddress();
+        if (address != null) {
+            isTerminalError(error, address.getResourceId(), ctx.getAuthenticatedDevice(), span.context())
+                    .onSuccess(isTerminalError -> {
+                        if (isTerminalError) {
+                            span.log("closing connection to device");
+                            conn.close();
+                            conn.disconnect();
+                        }
+                    })
+                    .onComplete(o -> span.finish());
+        }
```

Comment:
The `span` is not getting closed here in the (theoretical) case of the address being null. So, it would look better to add a corresponding "else" block here or return a Future here and do the closing where the method gets called.